Code review

MoerAI opened the pull request:
<!-- CURSOR_SUMMARY -->
> [!NOTE]
> Adds Riverpod-powered cart with product model, cart page, badge/controls, and add-to-cart animation; introduces async HTTP sample and minor refactors.
> 
> - **Cart & State Management (Riverpod)**:
>   - Add `lib/cart_provider.dart` with `CartItem`, list extensions (`itemCount`, `totalAmount`), and `CartNotifier` (add/remove/clear).
>   - Wire provider in `lib/main.dart` via `ProviderScope` and cart badge/actions.
> - **UI**:
>   - New product list in `lib/main.dart` using `flutter_card_swiper`, tap to `lib/pages/product_detail_page.dart`.
>   - Add `lib/pages/cart_page.dart` with quantity controls, delete, and total.
>   - Add animated feedback `lib/widgets/add_to_cart_animation.dart` triggered on add.
> - **Models**:
>   - Introduce `lib/models/product.dart` (`Equatable`, immutable, `copyWith`).
> - **Samples & Refactors**:
>   - Add `lib/async_sample.dart` demonstrating an HTTP GET before running a demo app.
>   - Make fields/constructors `final/const` in `lib/hero_todo.dart`.
>   - Remove legacy `lib/card_shop.dart`.
> - **Dependencies**:
>   - Update `pubspec.yaml` to add `http`, `flutter_riverpod`, `freezed_annotation`, `equatable` and dev tools (`build_runner`, `freezed`, `json_annotation`).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 15cec4c3525a91b9bb5eb1ed4915931ad659f35a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

atawLee committed:
async update

MoerAI committed:
flutter icn study task code review result

1. Map → List 변경:
  - 빠른 조회보다는 순서와 인덱스 기반 접근이 더 적합
  - UI 렌더링 시 toList() 변환 불필요
2. 단일 Provider 사용:
  - Provider 간 의존성 제거
  - Consumer에서 필요한 값만 계산하여 사용
3. Extension 활용:
  - 복잡한 계산 로직을 깔끔하게 캡슐화
  - reduce 함수로 총합 계산 최적화
4. 파일 분리:
  - 각 파일이 명확한 역할 분담
  - 코드 재사용성 향상

MoerAI committed:
몇가지 이슈 해결

1. 삭제 된 Main 함수 추가
2. Git 충돌로 상태와 코드 편집이 섞인 이슈 해결

marcuskim-braceletstudio commented on lib/models/product.dart:
equatable 또는 Freezed 적용필요

MoerAI replied:
Resolved. Product already extends Equatable on the current branch, and commit 45ecb30 applies the same value semantics to CartItem. Mutation proof failed identity/list equality on old head and all 7 cart-focused tests now pass.

marcuskim-braceletstudio commented on lib/hero_todo.dart:
class별 파일 분리 필요

MoerAI replied:
Reviewed, but not changed in this patch. hero_todo.dart is a small self-contained tutorial example and splitting each demo class would add navigation overhead without a behavioral boundary. The new production fixes and tests are isolated in their own files; I can split the tutorial only if the upstream maintainer wants that convention applied repo-wide.

marcuskim-braceletstudio commented on lib/async_sample.dart:
cpuBoundIsolate은 어디에 쓰는건가요?
isolate을 쓰시려면 이해를 하시고 쓰셔야 합니다.
AI가 짜주는 코드에 대해서 의구심및 모르는게 있을때 어떻게 돌아가는지는 꼭 확인부탁드려요
나중에 독이 될 수 있습니다

MoerAI replied:
Acknowledged. The unexplained cpuBoundIsolate code is no longer present on the branch. async_sample.dart now demonstrates a direct awaited HTTP request with explicit success/failure handling, which matches the learning goal without an isolate abstraction I cannot justify here.

marcuskim-braceletstudio commented on lib/cart_provider.dart:
CartItem Equatable 또는 Freezed 적용필요

MoerAI committed:
코드리뷰 반영 완료

1. Product.dart - Equatable 적용 완료
- Product class 상태 비교나 불변성 관리 필요
- Freezed을 사용하지 않은 이유는 Equatable이 조금 더 직관적이며 코드 수정이 적음

2. async_sample.dart - 불필요한 Isolate 코드 제거
- 복잡하지 않은 계산에 Isolate 사용
- Isolate 생성/통신 오버해드가 실제 계산 시간보다 커지는 성능 역효과 발생
- Future.delayed()는 CPU 바운드가 아닌 IO 바운드
- HTTP 요청 예시로 대체하고 에러 처리를 추가함

3. hero_todo.dart - Class의 분리는 필요하지 않다고 판단
- Class를 나누기에는 예제 규모가 너무 작음
- 높은 응집도와 낮은 재사용성

MoerAI commented on lib/main.dart:
```diff
@@ -0,0 +1,296 @@
+  void _showAddToCartAnimation(BuildContext context, Product product) {
+    final RenderBox buttonRenderBox = context.findRenderObject() as RenderBox;
+    final buttonPosition = buttonRenderBox.localToGlobal(Offset.zero);
+
```
Fixed in 45ecb30. The pressed button BuildContext now supplies the RenderBox and localToGlobal origin. Old-head mutation proof produced Offset(0.0, 56.0) versus the button at Offset(219.6, 707.0); the focused widget test now passes within 1px and the overlay completes without a pending timer.

MoerAI commented on lib/cart_provider.dart:
```diff
@@ -0,0 +1,111 @@
+      color: color ?? this.color,
+    );
+  }
+}
```
Fixed in 45ecb30. CartItem now extends Equatable with all value fields in props. Old-head mutation proof failed identical-item equality, list containment, and rebuilt-cart equality; the 7 cart tests and full 12-test suite now pass.
